File: QF_Cpp/test/NsCalculatorD/autocode/KeyboardImpl.cpp
//-*- Mode: C++; -*-
//===========================================================================
// This software contains Caltech/JPL confidential information.
//
// Copyright 2009-2011, by the California Institute of Technology.
// ALL RIGHTS RESERVED. United States Government Sponsorship Acknowledged.
// Any commercial use must be negotiated with the Office of Technology
// Transfer at the California Institute of Technology.
//
// This software may be subject to US export control laws and
// regulations. By accepting this document, the user agrees to comply
// with all applicable U.S. export laws and regulations, including the
// International Traffic and Arms Regulations, 22 C.F.R. 120-130 and the
// Export Administration Regulations, 15 C.F.R. 730-744. User has the
// responsibility to obtain export licenses, or other export authority as
// may be required before exporting such information to foreign countries
// or providing access to foreign persons.
//===========================================================================
//
//       File: KeyboardImpl.cpp
// Created on: 30-Jul-2011 10:20:19
//     Author: scheng@jpl.nasa.gov
//
// This file was stubbed by the JPL StateChart Autocoders, which converts UML
// Statecharts, in XML, to a C++ variant of Miro Samek's Quantum Framework.
//===========================================================================
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <log_event.h>
#include <qf_port.h>
#include <qassert.h>
#include <KeyboardImpl.h>
#include <UserEventImpl.h>
#include <StatechartSignals.h>

/**
 * Default constructor.
 */
KeyboardImpl::KeyboardImpl () {
    strcpy(this->machineName, "Keyboard");

    AttributeMapper::init(this);
}

/**
 * Required virtual destructor.
 */
KeyboardImpl::~KeyboardImpl () {
    AttributeMapper::clean(this);
}

void KeyboardImpl::setQActive (QActive* active) {
    m_active = active;
}


////////////////////////////////////////////
// Action and guard implementation methods
////////////////////////////////////////////

void KeyboardImpl::clearAll () {
	QF::publish(Q_NEW(KeyEvent, Clear));
}

void KeyboardImpl::clearEntry () {
	QF::publish(Q_NEW(KeyEvent, ClearEntry));
}

void KeyboardImpl::powerOff () {
	QF::publish(Q_NEW(KeyEvent, PowerOff));
}

void KeyboardImpl::sendEquals () {
	KeyEvent* ke = Q_NEW(KeyEvent, Equals);
	ke->keyId = '=';
	QF::publish(ke);
}

void KeyboardImpl::sendKey (char arg1) {
	KeyEvent* ke;
	if (arg1 == '0') {
		ke = Q_NEW(KeyEvent, Digit_0);
	} else {
		ke = Q_NEW(KeyEvent, Digit_1_9);
	}
	ke->keyId = arg1;
	QF::publish(ke);
}

void KeyboardImpl::sendPoint () {
	KeyEvent* ke = Q_NEW(KeyEvent, Point);
	ke->keyId = '.';
	QF::publish(ke);
}

void KeyboardImpl::sendOperator (char arg1) {
	KeyEvent* ke = Q_NEW(KeyEvent, Oper);
	ke->keyId = arg1;
	QF::publish(ke);
}

File: QF_Py/test/Calculator/autocode/KeyboardImpl.py
#
# Copyright 2009, 2010 California Institute of Technology.
# ALL RIGHTS RESERVED. U.S. Government Sponsorship acknowledged.
#
"""
File: KeyboardImpl.py

Date Created:  15-Apr-2010 18:29:03
Created By:    scheng

Python custom-implementation class for functions referenced in
the Keyboard Statechart model.
"""
# Python imports here
import sys
import logging

# QF imports here
from qf import framework
from qf import event
from UserEventImpl import KeyEvent

# Module globals initialized here
LOGGER = logging.getLogger('KeyboardLogger')


def printf(format, *args):
    # sys.stdout.write(format % args)
    LOGGER.info(format % args)


class KeyboardImpl(object):
    """
    Keyboard state machine implementation object.
    """
    def __init__(self, machine_name="Keyboard"):
        """
        Constructor
        """
        self.__machine_name = machine_name
        self.fr = framework.QF.getInstance()
        self.activeObjects = self.fr.getActiveDict()
        #
        print "*Accessible: IMPL['Keyboard']"

    def set(self, attr, value=True):
        """
        Sets the boolean value of an attribute, default to True
        """
        if type(value) == bool:
            self.__dict__['_'+attr] = value
            print attr, "set to", value
        else:
            print 'Value must be a bool, either "True" or "False" (no quotes)!'

    def get(self, attrname):
        """
        Gets the boolean value of an attribute
        """
        return self.__dict__['_'+attrname]

    def clear(self, attrname):
        """
        Clears the value of an attribute to False
        """
        self.__dict__['_'+attrname] = False

    def clearAll(self, e):
        """
        Implementation Action method for clearAll().
        The triggering event is supplied as a dictionary object 'e', so member
        data "data1" of the original Event object is accessible as e['data1'] .
        """
        self.fr.publish(KeyEvent("Clear"))

    def clearEntry(self, e):
        """
        Implementation Action method for clearError().
        The triggering event is supplied as a dictionary object 'e', so member
        data "data1" of the original Event object is accessible as e['data1'] .
        """
        self.fr.publish(KeyEvent("ClearEntry"))

    def powerOff(self, e):
        """
        Implementation Action method for powerOff().
        The triggering event is supplied as a dictionary object 'e', so member
        data "data1" of the original Event object is accessible as e['data1'] .
        """
        self.fr.publish(KeyEvent("PowerOff"))

    def sendEquals(self, e):
        """
        Implementation Action method for sendEquals().
        The triggering event is supplied as a dictionary object 'e', so member
        data "data1" of the original Event object is accessible as e['data1'] .
        """
        self.fr.publish(KeyEvent("Equals", '='))

    def sendKey(self, e, arg1):
        """
        Implementation Action method for sendKey().
        The triggering event is supplied as a dictionary object 'e', so member
        data "data1" of the original Event object is accessible as e['data1'] .
        """
        if arg1 == '0':
            self.fr.publish(KeyEvent("Digit_0", arg1))
        else:
            self.fr.publish(KeyEvent("Digit_1_9", arg1))

    def sendPoint(self, e):
        """
        Implementation Action method for sendPoint().
        The triggering event is supplied as a dictionary object 'e', so member
        data "data1" of the original Event object is accessible as e['data1'] .
        """
        self.fr.publish(KeyEvent("Point", '.'))

    def sendOperator(self, e, arg1):
        """
        Implementation Action method for sendOperator().
        The triggering event is supplied as a dictionary object 'e', so member
        data "data1" of the original Event object is accessible as e['data1'] .
        """
        self.fr.publish(KeyEvent("Oper", arg1))
